h3ndrk/hulk
# Communication v3

## Introduced Changes

This PR reimplements the communication protocol but this time as a separate crate. The separation allows to start the server without a framework (preparation for usage in the next generation behavior simulator).

I would like to delegate the client implementation to someone else.

For the reviewer:

1. Read the bullet points below in the documentation ToDo
2. Read `messages.rs` which gives an overview of the protocol
3. Maybe follow the data flow from `acceptor.rs` to `receiver.rs` over `router.rs` and `provider.rs` and then `sender.rs`
4. The test cases give some insight into which use cases are supported

## ToDo / Known Issues

- [x] Test the communication server manually
- [ ] Convert Communication Design document into proper documentation, also document the protocol
  - Communication Design
    - Messages
      - Requests are always textual, binary requests are protocol errors
      - At least within textual stream, messages are in order
      - Each sent data frame contains all textual data of this frame and binary references
        - This allows to have the binary stream separate and only the order in the textual stream matters
      - When the client subscribes, the request's id is used to store the subscription, unsubscribe in the future, and receive updated data from the server
      - Clients must specify with which cycler instance they want to speak
    - Databases
      - Clients choose subscription IDs, they need to be prefixed with a unique client identifier (e.g. peer address)
      - These asynchronous tasks exist: `provider n--1 router 1--1 receiver/sender`
        - Single `router` demultiplexes to `provider`
        - Each connected cycler instance spawns a `provider` task
      - Each `provider` task stores subscriptions
      - All `provider` tasks and the `receiver` tasks can send requests to the `router`
      - The `provider` first registers itself at the `router`
      - `GetNext` and `Subscribe` ids are shared and need to be unique
- [x] Remove all `println!()`
- [x] Remove communication server in `tools/`
- [x] Integrate communication server into existing framework
- [ ] Implement new protocol in communication client **Delegated to someone else**
- [x] Execute all CI checks
- [x] Reimplement additional output subscription notifications

## Ideas for Next Iterations (Not This PR)

- Directly generate BTreeMap in SerializeHierarchy
- Try to get rid of `#[derive(Default)]` for most of the types
- Implement parameters and injections in communication

## How to Test

Since the client does not implement the new protocol currently, you need to manually use e.g. `websocat` or a webbrowser to connect to the WebSocket and send/receive messages. The communication server should be available once you start the framework e.g. in Webots.
